compiler/noirc_evaluator/src/ssa/ir/function.rs

```diff
@@ -192,6 +193,23 @@ impl Function {
         blocks
     }
 
+    /// Collects all the reachable blocks of this function and returns them in pre-order.
```

Comment:
Are you sure we don't already have somewhere something like this? 

In any case, it would be good to spell out what we mean by _pre-order_. The code below looked deceptively simple, so I looked up the [first definition I found](https://www.w3schools.com/dsa/dsa_algo_binarytrees_preorder.php). The depicted algorithm shows a Depth First Search traversal of a binary tree, so it's not directly applicable to our code graph, but the one here looks like Breadth First Search. 

Say we have a graph like this:
```
b0 - b1 ---------- b5
   \              /
     b2 - b3 - b4 
```
It looks to me like the code below will yield `[b0, b1, b2, b5, b3, b4]`. Since `b5` comes before some of its predecessors, I'm not sure it can be considered pre-order.

Maybe some tests would be helpful.

Reply:
You are right! Yes, I saw we had PostOrder, looked up graph traversals, a question about post and pre order showed up and I must have looked at it wrong and concluded this is pre-order. I think what I implemented is BFS (Breadth-first search). I wonder if the above code returns the same thing as what I implemented, though (because the logic of `PostOrder` seems a bit more complex than that).

Reply:
In the end the `simple_reachable_blocks_optimization` logic was good for this PR too. I think I had problems before understanding we need to use dominance to implement this. So now the order of blocks traversal doesn't matter much as long as it goes in a way that if we visit a block, then blocks dominated by it will come next (this is guaranteed by the current logic though it's not explicitly documented).